### ISSUE TITLE
contributing for the artifacts with external dependencies (recyclerview)

### ISSUE THREAD

**Issue by adelnizamutdinov** [zserge/anvil#29]
Hey, I wanted to ask about that recyclerview patch.

How do you see external dependencies added to the project?
I tried adding a `recyclerView {}` flavor and `recyclerViewCompile "com.android.support:recyclerview-v7:$supportLibrary"` dependency, but `BaseDsl` expects each flavor to implement `trikita.anvil.DSL`

I am thinking about using flavor dimentions but that might introduce an unnecessary complexity to the publishing process and also something like sdk19RecyclerView variant doesn't make much sense.

What do you think? Do you have any vision on how to solve this? I personally would go for a module for each released artifacts and a core, something like:

- core (sdk10)
- sdk15
- sdk19
- recycler-view

But I'm not really sure how to handle the attrs part

**Comment by zserge** [comment]
Sorry for a late reply.

Anvil itself contains two parts - core and DSL. Core is just an implementation of the rendering mechanism. It doesn't care about classes other than View and ViewGroup. DSL is a class with handwritten and generated bindings for views and methods from Android SDK. I decided to ship both core + DSL in one package because core itself it useless, and Android SDK DSL is required in every Anvil application anyway.

Now about the external components. There are android support libraries and there are popular custom views around the github (floating action button, sliding tabs etc etc). If we would make a separate build flavour of Anvil for each of them - it would be impossible to maintain.

So I suggest to create a separate github project for each external library you want to wrap. For the RecyclerView I would imagine to have a separate "dsl" class with `recyclerView()` method to create view instances (like `button()` or `textView() in the SDK DSL). The same DSL class might also contain some setter wrappers for the attributes specific to RecyclerView only. Also there would be a handy wrapper for the adapters capable of displaying `Anvil.Renderable` layouts inside the list child views.

Summing up. Wrappers for external components depend on Anvil core and might also use Anvil's "common" DSL generated for Android SDK classes and methods. Which means they depend on Anvil. Anvil, on the other hand does not depend on them, which makes a perfect sense to distribute them as separate libraries.

**Comment by zserge** [comment]
There are some potential problems:

- Which API level of Anvil to specify as a dependency? sdk10? sdk15? sdk19? It should be possible to specify Anvil as _optional_ dependency, so it won't be fetched with your library, but instead the user of your library would add Anvil to his project explicitly. Then you may use sdk10 as an optional dependency and let users decide if they want a higher API level or not.

- How to use Anvil private classes? I try to keep Anvil public interface as simple as possible. That's why Anvil.Mount is not a public class. Basically, public classes/methods are used by Anvil apps, package-private classes/methods are used by Anvil "extension" libraries and private things are used by Anvil only. So if your RecyclerView Adapter uses Anvil.Mount (I bet it does) - you should keep the adapter implementation in the `trikita.anvil` package. If your library package is `com.example.recyclerview` then you should make `trikita.anvil.RecyclerViewAdapterImpl` class and its child class - `com.example.recyclerview.Adapter`. The `Impl` class would be able to use Anvil "developer" classes, while your public API would be in your own package.

- Name conflicts. Anvil DSL already captured all the cool names (e.g. `adapter()` method). If you create your own DSL class with the same methods - it won't work when imported statically. I see two approaches here:

```
// Use SDK DSL as a static import, use other DSLs explicitly
RecyclerDSL.view(() -> {
  size(FILL, FILL);
  backgroundColor(...);
  RecyclerDSL.adapter(...);
});

// Or use both as static imports, but use `recycler` prefix for all your methods:
recyclerView(() -> {
  size(FILL, FILL);
  backgroundColor(...);
  recyclerAdapter(...);
});
```

Another option would be to remove `final` keywords from the SDK `DSL` class, so you could extend your DSL class from it. Your class would then wrap "overridden" SDK DSL methods. It's not as easy as it sounds and it still would be a problem if someone wanted to use 2 libraries, say, RecyclerView wrapper and DrawerLayout wrapper. Even though each of them wraps DSL methods - the user would have name conflicts anyway.

**Comment by zserge** [comment]
I've started the first external library for Anvil - a helper for the recycler view APIs:

https://github.com/trikita/anvil-recyclerview

The approach is as described above - a single entry point which is a namespace class "Recycler" with methods view() to create new view instance, "adapter" to set an adapter etc.

The library depends on anvil-sdk10 which is the lowest supported API level, but this dependency is optional, so your project can use sdk15 or sdk19 instead.

Closing this issue, please submit all further issues about the recycler view here: https://github.com/trikita/anvil-recyclerview/issues
